opennebula/resource_opennebula_service_template.go

```diff
@@ -0,0 +1,395 @@
+				},
+			},
+			"uid": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Computed:    true,
+				Description: "ID of the user that will own the Service Template",
+			},
+			"gid": {
+				Type:        schema.TypeInt,
+				Optional:    true,
+				Computed:    true,
+				Description: "ID of the group that will own the Service Template",
+			},
+			"uname": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "Name of the user that will own the Service Template",
+			},
+			"gname": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "Name of the group that will own the Service Template",
+			},
+		},
+	}
+}
+
+func resourceOpennebulaServiceTemplateCreate(d *schema.ResourceData, meta interface{}) error {
+	controller := meta.(*goca.Controller)
+
+	var err error
+
+	if _, ok := d.GetOk("name"); !ok {
+		return fmt.Errorf("The Name is mandatory for creating a service template.")
+	}
+
+	if _, ok := d.GetOk("template"); !ok {
+		return fmt.Errorf("The service template body is mandatory.")
+	}
+
+	// Marshall the json
+	stemplate := &srv_tmpl.ServiceTemplate{}
+	err = json.Unmarshal([]byte(d.Get("template").(string)), stemplate)
+	if err != nil {
+		return err
+	}
+
+	err = controller.STemplates().Create(stemplate)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%v", stemplate.ID))
+	stc := controller.STemplate(stemplate.ID)
+
+	// Set the permissions on the service template if it was defined,
+	// otherwise use the UMASK in OpenNebula
+	if perms, ok := d.GetOk("permissions"); ok {
+		err = stc.Chmod(permissionUnix(perms.(string)))
+		if err != nil {
+			log.Printf("[ERROR] template permissions change failed, error: %s", err)
+			return err
+		}
+	}
+
+	if _, ok := d.GetOkExists("gid"); d.Get("gname") != "" || ok {
+		err = changeServiceTemplateGroup(d, meta, stc)
+		if err != nil {
+			return err
+		}
+	}
+
+	if _, ok := d.GetOkExists("uid"); d.Get("uname") != "" || ok {
+		err = changeServiceTemplateOwner(d, meta, stc)
+		if err != nil {
+			return err
+		}
+	}
+
+	if d.Get("name") != "" {
+		err = changeServiceTemplateName(d, meta, stc)
+		if err != nil {
+			return err
+		}
+	}
+
+	return resourceOpennebulaServiceTemplateRead(d, meta)
+}
+
+func resourceOpennebulaServiceTemplateRead(d *schema.ResourceData, meta interface{}) error {
+	stc, err := getServiceTemplateController(d, meta)
+	if err != nil {
+		return err
+	}
+
+	st, err := stc.Info()
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%v", st.ID))
+	d.Set("name", st.Name)
+	d.Set("uid", st.UID)
+	d.Set("gid", st.GID)
+	d.Set("uname", st.UName)
+	d.Set("gname", st.GName)
+	err = d.Set("permissions", permissionsUnixString(*st.Permissions))
+	if err != nil {
+		return err
+	}
+
+	// Get service.Template as map
+	tmpl_byte, err := json.Marshal(st.Template)
+	if err != nil {
+		return err
+	}
+	d.Set("template", "{\"TEMPLATE\":"+string(tmpl_byte)+"}")
```

Comment:
As a future improvement it could be nice to manage the template directly with terraform like it is now the case for VM templates